ClassCastException in qute template

In a qute template, the following will trigger a ClassCastException:
```
    {#if !}

    {/if}
```
See:
```
[Trace - 18:50:59] Sending request 'textDocument/codeAction - (2217)'.
Mar 15, 2023 6:50:59 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at com.redhat.qute.services.codeactions.QuteCodeActionForUndefinedObject.doCodeActions(QuteCodeActionForUndefinedObject.java:77)
	at com.redhat.qute.services.QuteCodeActions.doCodeActions(QuteCodeActions.java:116)
	at com.redhat.qute.services.QuteLanguageService.doCodeActions(QuteLanguageService.java:104)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$codeAction$9(TemplateFileTextDocumentService.java:165)
	at com.redhat.qute.ls.commons.ModelTextDocuments.lambda$computeModelAsyncCompose$1(ModelTextDocuments.java:144)
	... 7 more

[Trace - 18:50:59] Received response 'textDocument/codeAction - (2217)' in 1ms. Request failed: Internal error. (-32603).
[Error - 18:50:59] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at com.redhat.qute.services.codeactions.QuteCodeActionForUndefinedObject.doCodeActions(QuteCodeActionForUndefinedObject.java:77)
	at com.redhat.qute.services.QuteCodeActions.doCodeActions(QuteCodeActions.java:116)
	at com.redhat.qute.services.QuteLanguageService.doCodeActions(QuteLanguageService.java:104)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$codeAction$9(TemplateFileTextDocumentService.java:165)
	at com.redhat.qute.ls.commons.ModelTextDocuments.lambda$computeModelAsyncCompose$1(ModelTextDocuments.java:144)
	... 7 more

[Trace - 18:50:59] Sending notification 'textDocument/didSave'.
[Trace - 18:51:00] Sending notification 'workspace/didChangeWatchedFiles'.
[Trace - 18:51:00] Received notification 'textDocument/publishDiagnostics'.
[Trace - 18:51:00] Sending request 'textDocument/codeAction - (2218)'.
Mar 15, 2023 6:51:00 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
java.util.concurrent.CompletionException: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at com.redhat.qute.services.codeactions.QuteCodeActionForUndefinedObject.doCodeActions(QuteCodeActionForUndefinedObject.java:77)
	at com.redhat.qute.services.QuteCodeActions.doCodeActions(QuteCodeActions.java:116)
	at com.redhat.qute.services.QuteLanguageService.doCodeActions(QuteLanguageService.java:104)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$codeAction$9(TemplateFileTextDocumentService.java:165)
	at com.redhat.qute.ls.commons.ModelTextDocuments.lambda$computeModelAsyncCompose$1(ModelTextDocuments.java:144)
	... 7 more

[Trace - 18:51:00] Received response 'textDocument/codeAction - (2218)' in 2ms. Request failed: Internal error. (-32603).
[Error - 18:51:00] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.redhat.qute.parser.expression.Parts cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.Parts and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at com.redhat.qute.services.codeactions.QuteCodeActionForUndefinedObject.doCodeActions(QuteCodeActionForUndefinedObject.java:77)
	at com.redhat.qute.services.QuteCodeActions.doCodeActions(QuteCodeActions.java:116)
	at com.redhat.qute.services.QuteLanguageService.doCodeActions(QuteLanguageService.java:104)
	at com.redhat.qute.ls.template.TemplateFileTextDocumentService.lambda$codeAction$9(TemplateFileTextDocumentService.java:165)
	at com.redhat.qute.ls.commons.ModelTextDocuments.lambda$computeModelAsyncCompose$1(ModelTextDocuments.java:144)
	... 7 more

```

Tested with the latest vscode-quarkus pre-release (v1.13.2023031404)

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/expression/ObjectPart.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/expression/ObjectPart.java
@@ -17,6 +17,7 @@ import com.redhat.qute.parser.template.Expression;
 import com.redhat.qute.parser.template.JavaTypeInfoProvider;
 import com.redhat.qute.parser.template.Parameter;
 import com.redhat.qute.parser.template.Section;
+import com.redhat.qute.parser.template.SectionKind;
 import com.redhat.qute.parser.template.Template;
 import com.redhat.qute.parser.template.sections.LoopSection;
 
@@ -33,7 +34,7 @@ import com.redhat.qute.parser.template.sections.LoopSection;
 
 public class ObjectPart extends Part {
 
-	private Boolean notComputed;
+	private int startName = -1;
 
 	public ObjectPart(int start, int end) {
 		super(start, end);
@@ -46,29 +47,23 @@ public class ObjectPart extends Part {
 
 	@Override
 	public int getStartName() {
-		computeNotIfNeeded();
-		return super.getStartName();
-	}
-
-	private void computeNotIfNeeded() {
-		if (notComputed != null) {
-			return;
+		if (startName != -1) {
+			return startName;
 		}
-		computeNot();
-	}
-
-	private synchronized void computeNot() {
-		if (notComputed != null) {
-			return;
+		startName = super.getStartName();
+		Parameter parameter = getOwnerParameter();
+		if (parameter != null) {
+			Section section = parameter.getOwnerSection();
+			if (section != null && section.getSectionKind() == SectionKind.IF) {
+				String text = getOwnerTemplate().getText();
+				if (text.charAt(startName) == '!') {
+					// ex : !true
+					// !true --> true
+					startName++;
+				}
+			}
 		}
-		String text = getOwnerTemplate().getText();
-		int start = super.getStartName();
-		if (text.charAt(start) == '!') {
-			// ex : !true
-			// !true --> true
-			super.setStart(start + 1);
-		}
-		notComputed = Boolean.TRUE;
+		return startName;
 	}
 
 	public JavaTypeInfoProvider resolveJavaType() {

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java
@@ -707,7 +707,15 @@ class QuteDiagnostics {
 				return null;
 			}
 
-			String literalJavaType = LiteralSupport.getLiteralJavaType(objectPart.getPartName());
+			String partName = objectPart.getPartName();
+			if (partName.isEmpty()) {
+				// This case comes from when only ! is used in #if
+				// ex : {#if !}
+				// In this case ! is not considered as an object part.
+				return null;
+			}
+
+			String literalJavaType = LiteralSupport.getLiteralJavaType(partName);
 			if (literalJavaType != null) {
 				// The object part is a literal type (ex : true)
 				return null;
@@ -723,7 +731,7 @@ class QuteDiagnostics {
 				return null;
 			}
 
-			if (CaseSection.isCaseSection(ownerSection)) {
+			if (Section.isCaseSection(ownerSection)) {
 				// Skip validation for case section, is done later
 				return null;
 			}

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsForSyntax.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsForSyntax.java
@@ -28,6 +28,8 @@ import com.redhat.qute.project.tags.UserTag;
 
 import io.quarkus.qute.Engine;
 import io.quarkus.qute.EngineBuilder;
+import io.quarkus.qute.ErrorCode;
+import io.quarkus.qute.ParserError;
 import io.quarkus.qute.TemplateException;
 import io.quarkus.qute.TemplateNode.Origin;
 import io.quarkus.qute.UserTagSectionHelper;
@@ -67,17 +69,29 @@ public class QuteDiagnosticsForSyntax {
 			Engine engine = engineBuilder.build();
 			engine.parse(templateContent);
 		} catch (TemplateException e) {
-			String message = e.getMessage();
-			if (message.contains("no section helper found for")) {
+			if (isIgnoreError(e)) {
 				// Ignore error "no section helper found for" which is managed with
 				// QuteDiagnostic to highlight the section start correctly.
 				return;
 			}
+			String message = e.getMessage();
 			Range range = createRange(e, template);
 			Diagnostic diagnostic = createDiagnostic(range, message, DiagnosticSeverity.Error,
 					QuteErrorCode.SyntaxError);
 			diagnostics.add(diagnostic);
 		}
+	}
+
+	private static boolean isIgnoreError(TemplateException e) {
+		// As the Qute real parser is not fault tolerant (it report just one error) and
+		// as it reports bad offset for the error
+		// we ignore some errors which are managed by QuteDiagnostics to highlight
+		// several errors and highlight section, object part, etc correctly.
+		ErrorCode code = e.getCode();
+		if (ParserError.NO_SECTION_HELPER_FOUND == code) {
+			return true;
+		}
+		return false;
 	}
 
 	private static void addUserTag(Collection<UserTag> tags, EngineBuilder engineBuilder) {

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithIfSectionTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithIfSectionTest.java
@@ -121,6 +121,14 @@ public class QuteDiagnosticsInExpressionWithIfSectionTest {
 	}
 
 	@Test
+	public void onlyNotOperator() {
+		String template = "{#if !}{/if}";
+		Diagnostic d = d(0, 0, 0, 0, QuteErrorCode.SyntaxError, "Unsupported param type: NOT",
+				DiagnosticSeverity.Error);
+		testDiagnosticsFor(template, d);
+	}
+
+	@Test
 	public void noOptionalParameterInIfBlock() {
 		String template = "{#if foo}\r\n" + //
 				"	{foo}\r\n" + //

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithLetSectionTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/diagnostics/QuteDiagnosticsInExpressionWithLetSectionTest.java
@@ -106,4 +106,11 @@ public class QuteDiagnosticsInExpressionWithLetSectionTest {
 						ConfigurationItemEditType.update, "ignore", //
 						d)));
 	}
+
+	@Test
+	public void notAsObjectPartName() throws Exception {
+		String template = "{@int !value}\r\n" + //
+				"{#let name=!value /}";
+		testDiagnosticsFor(template);
+	}
 }

